packages/turf-constrained-tin/index.js

```diff
@@ -0,0 +1,897 @@
+ * @returns {FeatureCollection<Polygon>} TIN output
+ * @example
+ * // generate some random point data
+ * var points = turf.randomPoint(30, {bbox: [50, 30, 70, 50]});
+ *
+ * // add a random property to each point between 0 and 9
+ * for (var i = 0; i < points.features.length; i++) {
+ *   points.features[i].properties.z = ~~(Math.random() * 9);
+ * }
+ * var tin = turf.constrainedTin(points, [], 'z');
+ *
+ * //addToMap
+ * var addToMap = [tin, points]
+ * for (var i = 0; i < tin.features.length; i++) {
+ *   var properties  = tin.features[i].properties;
+ *   properties.fill = '#' + properties.a + properties.b + properties.c;
+ * }
+ */
+module.exports = function(points, edges, z) {
+    var isPointZ = false;
+    var ret = {
+        vert: points.features.map(function(point) {
+            var xy = point.geometry.coordinates;
+            return new Point(xy[0], xy[1]);
+        }),
+        z: points.features.map(function(point) {
+            if (z) {
+                return point.properties[z];
+            } else if (point.geometry.coordinates.length === 3) {
+                isPointZ = true;
+                return point.geometry.coordinates[2];
+            }
+        })
+    };
+    loadEdges(ret, edges)
+    delaunay(ret);
+    constrainEdges(ret);
+    var keys = ['a', 'b', 'c'];
+    return helper.featureCollection(ret.tri.map(function(indices) {
+        var properties = {};
+        var coords = indices.map(function(index, i) {
+            var coord = [ret.vert[index].x, ret.vert[index].y];
+            if (ret.z[index] !== undefined) {
+                if (isPointZ) {
+                    coord[2] = ret.z[index];
+                } else {
+                    properties[keys[i]] = ret.z[index];
+                }
+            }
+            return coord; 
+        })
+        coords[3] = coords[0];
+        return helper.polygon([coords], properties);
+    }));
+};
+
+var Point = function(x, y) {
+    this.x = x;
+    this.y = y;
+};
+Point.prototype.dot = function(p1) {
+    return (this.x*p1.x + this.y*p1.y);
+};
+Point.prototype.add = function(p1) {
+    return new Point(this.x + p1.x, this.y + p1.y);
+};
+Point.prototype.sub = function(p1) {
+    return new Point(this.x - p1.x, this.y - p1.y);
+};
+Point.prototype.scale = function(s) {
+    return new Point(this.x*s, this.y*s);
+};
+Point.prototype.sqDistanceTo = function(p1) {
+    return (this.x - p1.x)*(this.x - p1.x) + (this.y - p1.y)*(this.y - p1.y);
+};
+Point.prototype.toStr = function() {
+    return "(" + this.x.toFixed(3) + ", " + this.y.toFixed(3) + ")";
+};
+Point.prototype.copyFrom = function(p) {
+    this.x = p.x;
+    this.y = p.y;
+};
+
+function cross(vec0, vec1)
+{
+    return (vec0.x*vec1.y - vec0.y*vec1.x);
+}
+
+function getPointOrientation(edge, p)
+{
+    var vec_edge01 = edge[1].sub(edge[0]);
+    var vec_edge0_to_p = p.sub(edge[0]);
+    return cross(vec_edge01, vec_edge0_to_p);
+}
+
+//Some variables for rendering
```

Comment:
Maybe this is part of the library, but I'm not sure I understand what these apparently arbitrary values represent. Particularly the screenL(ength?) 🤔

Reply:
It is coming from the original implementation.
https://github.com/savithru-j/cdt-js/blob/master/cdt.js

The original implementation did not separate the UI from the computation logic.
So I made many efforts to remove UI-related variables and functions but only 
* min_coord
* screenL
* boundingL
these 3 variables are used in both UI and computation logic.
So they are kept with their original names.

Rough my understanding is,
* min_coord - coordinate of the origin point of the field which all points must be on it.
* screenL - Width/height of the field.
* boundingL - Far away point x/y values used for the initial triangle.

So, if it should be, I will change these variable names as follow:
* min_coord -> field_origin
* screenL -> fieldSize
* boundingL -> boundingTriangleSize

Reply:
Done